tests/test_shell.py

```diff
@@ -199,13 +324,234 @@ def test_remove(self):
+        assert len(project_b) == 6
+
+    def test_sync_document(self):
+        self.call('python -m signac init ProjectA'.split())
+        project_a = signac.Project()
+        project_b = signac.init_project('ProjectB', os.path.join(self.tmpdir.name, 'b'))
+        job_src = project_a.open_job({'a': 0})
+        job_dst = project_b.open_job({'a': 0})
+
+        def reset():
+            job_src.document['a'] = 0
+            job_src.document['nested'] = dict(a=1)
+            job_dst.document['a'] = 1
+            job_dst.document['nested'] = dict(a=2)
+
+        reset()
+        assert job_dst.document != job_src.document
+        with pytest.raises(ExitCodeError):
+            self.call('python -m signac sync {} {}'
+                      .format(os.path.join(self.tmpdir.name, 'b'), self.tmpdir.name).split())
+        assert job_dst.document != job_src.document
+
+        self.call('python -m signac sync {} {} --no-key'
+                  .format(os.path.join(self.tmpdir.name, 'b'), self.tmpdir.name).split())
+        assert job_dst.document != job_src.document
+        assert job_dst.document['a'] != job_src.document['a']
+        assert job_dst.document['nested'] != job_src.document['nested']
+
+        reset()     # only sync a
+        self.call('python -m signac sync {} {} --key a'
+                  .format(os.path.join(self.tmpdir.name, 'b'), self.tmpdir.name).split())
+        assert job_dst.document != job_src.document
+        assert job_dst.document['nested'] != job_src.document['nested']
+        assert job_dst.document['a'] == job_src.document['a']
+
+        reset()     # only sync nested
+        self.call('python -m signac sync {} {} --key nested'
+                  .format(os.path.join(self.tmpdir.name, 'b'), self.tmpdir.name).split())
+        assert job_dst.document != job_src.document
+        assert job_dst.document['a'] != job_src.document['a']
+        assert job_dst.document['nested'] == job_src.document['nested']
+
+        reset()
+        self.call('python -m signac sync {} {} --all-key'
+                  .format(os.path.join(self.tmpdir.name, 'b'), self.tmpdir.name).split())
+        assert job_dst.document == job_src.document
+        assert job_dst.document['nested'] == job_src.document['nested']
+        assert job_dst.document['a'] == job_src.document['a']
+
+        with pytest.raises(ExitCodeError):
+            self.call('python -m signac sync {} {} --all-key --no-key'
+                      .format(os.path.join(self.tmpdir.name, 'b'), self.tmpdir.name).split())
+
+    def test_sync_file(self):
+        self.call('python -m signac init ProjectA'.split())
+        project_a = signac.Project()
+        project_b = signac.init_project('ProjectB', os.path.join(self.tmpdir.name, 'b'))
+        job_src = project_a.open_job({'a': 0})
+        job_dst = project_b.open_job({'a': 0})
+        for i, job in enumerate([job_src, job_dst]):
+            with job:
+                with open('test', 'w') as file:
```

Comment:
Could be simplified so that `with job` isn't needed: `with open(job.fn('test'), 'w') as file:`